WalletWasabi/Crypto/ZeroKnowledge/Verifier.cs

```diff
@@ -0,0 +1,47 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using WalletWasabi.Crypto.Groups;
+using WalletWasabi.Helpers;
+
+namespace WalletWasabi.Crypto.ZeroKnowledge
+{
+	public static class Verifier
+	{
+		public static bool Verify(KnowledgeOfRepresentation proof, GroupElement publicPoint, IEnumerable<GroupElement> generators)
+		{
+			Guard.False($"{nameof(publicPoint)}.{nameof(publicPoint.IsInfinity)}", publicPoint.IsInfinity);
+			Guard.NotNullOrEmpty(nameof(generators), generators);
+
+			if (publicPoint == proof.Nonce)
+			{
+				throw new InvalidOperationException($"{nameof(publicPoint)} and {nameof(proof.Nonce)} should not be equal.");
+			}
+
+			foreach (var generator in generators)
+			{
+				Guard.False($"{nameof(generator)}.{nameof(generator.IsInfinity)}", generator.IsInfinity);
+			}
+
+			var nonce = proof.Nonce;
+			var responses = proof.Responses;
+
+			var challenge = Challenge.Build(publicPoint, nonce, generators);
+			var a = challenge * publicPoint + nonce;
+
+			var b = GroupElement.Infinity;
+			foreach (var (response, generator) in responses.ZipForceEqualLength(generators))
+			{
+				b += response * generator;
+			}
+			return a == b;
+		}
+
+		public static bool Verify(KnowledgeOfDiscreteLog proof, GroupElement publicPoint, GroupElement generator)
```

Comment:
same question re the additional type when this is just a special case of KnowledgeOfRepresentation

if it's in order to use the typesystem to avoid mixups between the specific proof types/structures used in WabiSabi i think it makes more sense to do that at a slightly higher level (e.g. NullCredentialProof)

Reply:
It's not a separate type.

```cs
public class KnowledgeOfDiscreteLog : KnowledgeOfRepresentation
{
	public KnowledgeOfDiscreteLog(GroupElement nonce, Scalar response)
		: base(nonce, new[] { response })
	{
	}
		public Scalar Response => Responses.First();
}
```

Reply:
well, it's a subtype but it's still a separate type... why do we need both?